experimental/dds/tree2/src/feature-libraries/sequence-field/compose.ts

```diff
@@ -227,62 +230,45 @@ function composeMarks<TNodeChange>(
 			);
 		}
 
-		return withRevision(withNodeChange(newMark, nodeChange), newRev);
+		return withRevision(normalizeTransient(newTransient, nodeChange), newRev);
 	}
-	if (isTransientEffect(baseMark)) {
+	if (isTransientEffect(baseMark) || isDetachOfRemovedNodes(baseMark)) {
+		const baseTransient = asTransient(baseMark);
 		if (markFillsCells(newMark)) {
-			if (isMoveDestination(baseMark.attach) && isMoveSource(baseMark.detach)) {
+			if (isMoveDestination(baseTransient.attach) && isMoveSource(baseTransient.detach)) {
 				assert(isMoveDestination(newMark), "Unexpected mark type");
 				setEndpoint(
 					moveEffects,
 					CrossFieldTarget.Source,
 					getEndpoint(newMark, newRev),
-					baseMark.count,
-					{ revision: baseMark.attach.revision, localId: baseMark.attach.id },
+					baseTransient.count,
+					{ revision: baseTransient.attach.revision, localId: baseTransient.attach.id },
 				);
 			}
 
 			const originalAttach = withRevision(
 				withNodeChange(
-					{ ...baseMark.attach, cellId: baseMark.cellId, count: baseMark.count },
+					{
+						...baseTransient.attach,
+						cellId: baseTransient.cellId,
+						count: baseTransient.count,
+					},
 					nodeChange,
 				),
-				baseMark.attach.revision,
+				baseTransient.attach.revision,
 			);
-
-			// TODO: This assumes that the original attach was successful.
-			// We should probably treat a muted transient effect as a noop.
 			return originalAttach;
-		}
-		// Noop and Placeholder marks must be muted because the node they target has been deleted.
-		// Detach marks must be muted because the cell is empty.
-		if (newMark.type === NoopMarkType || newMark.type === "Placeholder" || isDetach(newMark)) {
+		} else if (newMark.type === NoopMarkType || newMark.type === "Placeholder") {
 			assert(
 				newMark.cellId !== undefined,
 				0x718 /* Invalid node-targeting mark after transient */,
 			);
+			return withNodeChange(baseMark, nodeChange);
+		} else {
+			// This case is handled above as a transient.
+			assert(!isDetach(newMark), "Unexpected mark type");
```

Comment:
A comment here about what kinds of mark `newMark` could be would be nice. We've checked that it's not a transient, active attach, or detach (since they're effectively transient). So it must be an inactive attach, right? Do we ever have inactive attaches with the new merge semantics?

Reply:
I suppose this is also where we would handle inactive detaches, if we had any. Even if we had inactive attaches and detaches, I think this block could be merged with the previous else-if block, because an inactive mark should behave the same as a noop here.